### PR TITLE
Add glommio support

### PR DESCRIPTION
**Pull request by qiujiangkun**
This PR add basic support for glommio current thread executor. thread pool version is not available since glommio official only have limited support for it

### CODE REVIEW

**Comment by najamelan** [comment]
Hi, thanks for the contribution. I will have a look tomorrow.

**Comment by qiujiangkun** [comment]
I've added a work-stealing algorithm for the thread pool. In fact, enhancing this lib with glommio is part of my job. It gives me a good framework to work on, so I decide to post my modifications it here. Maybe further changes will be too customized to be merged here.

**Comment by qiujiangkun** [comment]
> I have started reviewing. One big thing is I am happy to include the wrapper for the single threaded executor, but I don't think this crate is the right place to implement a multi-threaded executor. If it doesn't exist, I think it can be proposed to glommio or if not put in a separate crate. We can then wrap it, or it can just implement the traits directly in it's own crate.

Sure. There is an official thread pool implementation but it is not useful. There are discussions about distributing tasks across multiple threads in glommio, but not decided yet. We can wait until then.

> I wonder how useful it is as it seems its main interest is the combination with io_uring, however once it's behind a trait, it's no longer known that it is a glommio executor, so a library taking in impl Spawn won't be able to benefit from glommio IO. So apart from the IO, does glommio really bring something new to the table in terms of executor? I can see it can be handy for an application that is already using glommio and has to pass an executor to a lib, so for me that's reason enough to include it.

My main use of it is not `io_uring`, but its thread model. As far as I know, it is the only mature runtime that supports `pin_to_cpu`. It beat tokio by 30% in one of my benchmarks.

> I would prefer a single struct GlommioCt rather than a builder. I don't really get what the builder adds in value. The constructor of GlommioCt can just take cpu: Option<usize>, name: &str. That keeps the API simpler.

> It should be declared !Send, as its a single threaded executor and we would initialize it in the constructor, so it shouldn't be allowed to be sent across threads. So it needs to hold PhantomData<Rc<()>>.

I found TokioCt holds `Arc`(shouldn't it be `Rc` since it is `!Send`?)
```rust
pub struct TokioCt {
    pub(crate) exec: Arc<Runtime>,
    pub(crate) local: Arc<LocalSet>,
}
```

There are some problems:

- There is `LocalExecutorBuilder` that holds many configuration items and I was mimicking `TokioCtBuilder`
- I was avoiding some of the borrow-checking issues without help of Rc(In the examples/glommio.rs). 
- There are heavy side-affects when you build a `LocalExecutor` and I don't think it should happen in a constructor. Maybe we should hold Rc<LocalExecutor> in GlommioCt? But LocalExecutor does not expose anything other than `id` and `run`. It uses `scoped-tls` internally
> If possible, for the JoinHandle it should wrap the one from gommio instead of using RemoteHandle, as that last one has overhead.

I should have seen `glommio::JoinHandle` in the beginning.

> The code currently throws clippy warnings.

> It should have integration tests. It's possible to just copy/paste the ones from LocalPool and replace that with Glommio.

**Comment by najamelan** [comment]
> My main use of it is not `io_uring`, but its thread model. As far as I know, it is the only mature runtime that supports `pin_to_cpu`. It beat tokio by 30% in one of my benchmarks.

Fair enough, didn't think of that.

> Sure. There is an official thread pool implementation but it is not useful. There are discussions about distributing tasks across multiple threads in glommio, but not decided yet. We can wait until then.

Yes, helping them fix it is a much better strategy than rolling something custom here.

> I found TokioCt holds `Arc`(shouldn't it be `Rc` since it is `!Send`?)

Yes, absolutely. I also noticed that while looking through the code earlier on.

> There are some problems:
> 
>     * There is `LocalExecutorBuilder` that holds many configuration items and I was mimicking `TokioCtBuilder`

So with tokio, I had to introduce (unfortunately) a builder for tokio 1 because before I just took `tokio::runtime::Builder` in the constructor. That allowed users to set their own configuration and then I would set it to single or multithreaded depending on which type they were instantiating. However the new API, you have to decide whether it's multi or single to create `tokio::runtime::Builder` and afterwards there is no way of finding out which one it is, nor change it. That means people could have constructed `TokioCt` with a multithreaded tokio runtime or `TokioTp` with a singlethreaded one. Since I want to avoid such accidents and the types in async_executor guarantee what kind you get, it was the only way, eg we construct the tokio builder ourselves and then borrow it to the user if they want to configure it.

So glommio doesn't have this problem, so we don't need a builder I think.

> 
>     * I was avoiding some of the borrow-checking issues without help of Rc(In the examples/glommio.rs).

Yes a usage like in the example I like, but then why have `GlommioCt`? It seems to just use `glommio_crate::Task` to spawn. I would put that in the trait impls for the `GlommioCtBuilder` and rename it to remove the `Builder` part.

>     * There are heavy side-affects when you build a `LocalExecutor` and I don't think it should happen in a constructor. Maybe we should hold Rc in GlommioCt? But LocalExecutor does not expose anything other than `id` and `run`. It uses `scoped-tls` internally

Yes, it looks like it is global, so no need to hold it. However, is it ok to call it twice in the same thread? If not, how do we prevent that? Also what are the side effects you are worried about? If it has to be called at least once before using the executor, then why not in the constructor. Or is it possible to use the executor without ever calling it?

**Comment by qiujiangkun** [comment]
> Yes, it looks like it is global, so no need to hold it. However, is it ok to call it twice in the same thread? If not, how do we prevent that? Also what are the side effects you are worried about? If it has to be called at least once before using the executor, then why not in the constructor. Or is it possible to use the executor without ever calling it?

It is not possible to call constructor twice with a single `LocalExecutor`. It is Okay to initialize it twice, since scoped-tls will behave normally. But it may disturbe the thread's cpu affinity. We can't do much about it.

File: Cargo.yml [hunk 1]
```diff
@@ -118,12 +122,12 @@ dependencies:
   async_std_crate       : { version: ^1.6  , optional: true, package: async-std, features: [ unstable ] }
   tokio                 : { version: ^1    , optional: true }
   pin-utils             : { version: ^0.1  , optional: true }
-
+  crossbeam             : { version: ^0.8  , optional: true }
```
Comment (najamelan) [inline Cargo.yml:125]: Do we still need deps on crossbeam and futures-executor?

File: README.md [hunk 2]
```diff
@@ -22,10 +22,11 @@ The currently supported executors are (file an issue on GitHub if you want to se
 - [async-std](https://docs.rs/async-std) - supports spawning `!Send` futures and works on Wasm (uses async-global-executor and bindgen under the hood).
 - [tokio](https://docs.rs/tokio) CurrentThread - [`tokio::runtime::Runtime`] with basic scheduler and a LocalSet. (supports spawning `!Send` futures)
 - [tokio](https://docs.rs/tokio) ThreadPool - [`tokio::runtime::Runtime`] with threadpool scheduler.
+- [glommio](https://docs.rs/glommio) Glommio is a Cooperative Thread-per-Core crate for Rust & Linux based on io_uring. Includes both current_thread and thread_pool implementation
```
Comment (najamelan) [inline README.md:25]: So maybe we leave out multithread for now.

File: src/glommio_ct.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+use futures_task::{FutureObj, LocalSpawn, Spawn, SpawnError};
+use futures_util::future::LocalFutureObj;
+use glommio_crate::{LocalExecutor, LocalExecutorBuilder};
+use std::future::Future;
+use std::rc::Rc;
+use crate::{LocalSpawnHandle, SpawnHandle, InnerJh, JoinHandle};
+
+/// A simple glommio runtime builder
+#[derive(Debug)]
+pub struct GlommioCt {
```
Comment (najamelan) [inline src/glommio_ct.rs:10]: I feel the API is much nicer. Just wondering, what happens if someone creates 2 of these on the same thread? Should we prevent it?

File: src/glommio_ct.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+impl LocalSpawn for GlommioCt {
+    fn spawn_local_obj(&self, future: LocalFutureObj<'static, ()>) -> Result<(), SpawnError> {
+        glommio_crate::Task::local(future).detach();
+        Ok(())
+    }
+}
+
+
+impl<Out: Send + 'static> LocalSpawnHandle<Out> for GlommioCt {
```
Comment (najamelan) [inline src/glommio_ct.rs:41]: I see on `TokioCt` the `Out` type doesn't have a `Send` bound. This should work without it.

**Review by najamelan — CHANGES_REQUESTED** [review]
It starts to look pretty good. Do you want to add integration tests to see if it all works? You can copy paste from `LocalPool` and replace the executor.

File: src/glommio_ct.rs (hunk 3 above)
Comment (qiujiangkun) [reply src/glommio_ct.rs:10]: I think we can leave it as is for now, since if user want to create two, they have to explicitly call `GlommioCt::new(name, cpu)`

**Comment by qiujiangkun** [comment]
There is one bug: the example in README.md does not compile(`cargo test` tries)

File: Cargo.yml (hunk 1 above)
Comment (qiujiangkun) [reply Cargo.yml]: I needed for GlommioTp. Now I removed it

File: tests/glommio_ct.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
+#![ cfg( feature = "glommio" ) ]
+
+// ✔ pass a     ThreadPool  to a function that takes exec: `impl SpawnHandle`
```
Comment (najamelan) [inline tests/glommio_ct.rs:3]: Looks like you took the tests from `ThreadPool`. This does not test any of the `LocalSpawn` and `LocalSpawnHandle` functionality. Actually the right one to copy from is `TokioCt`. I suggested `LocalPool` before because it doesn't require a builder, but it has less tests and the builder actually is only one line of code, so it will be the same to adapt.

File: src/local_spawn_handle.rs [hunk 6]
```diff
@@ -20,7 +20,7 @@ pub trait LocalSpawnHandle<Out: 'static>
 }
 
 
-/// Let's you spawn a !Send future and get a [JoinHandle] to await the output of a future.
+/// Let you spawn a !Send future and get a [JoinHandle] to await the output of a future.
```
Comment (najamelan) [inline src/local_spawn_handle.rs:23]: Actually the correct one is "Lets".

**Review by najamelan — CHANGES_REQUESTED** [review]
There seems to be also an issue with examples/spawn_handle_multi.rs. It looks like the `JoinHandle` now has `NonNull` in it which isn't `Send`. Not sure yet what the best solution is.

File: src/local_spawn_handle.rs (hunk 6 above)
Comment (qiujiangkun) [reply src/local_spawn_handle.rs:23]: > There seems to be also an issue with examples/spawn_handle_multi.rs. It looks like the `JoinHandle` now has `NonNull` in it which isn't `Send`. Not sure yet what the best solution is.


Yes I also find out, which is bad

```rust
error[E0277]: `NonNull<()>` cannot be sent between threads safely
   --> async_executors/src/core/join_handle.rs:262:2
    |
262 |     static_assertions::assert_impl_all!(JoinHandle<()>: Send);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |     |
    |     `NonNull<()>` cannot be sent between threads safely
    |     required by this bound in `assert_impl_all`
    |
    = help: within `join_handle::JoinHandle<()>`, the trait `std::marker::Send` is not implemented for `NonNull<()>`
    = note: required because it appears within the type `glommio::task::JoinHandle<()>`
    = note: required because it appears within the type `Option<glommio::task::JoinHandle<()>>`
    = note: required because it appears within the type `glommio::executor::multitask::Task<()>`
    = note: required because it appears within the type `glommio::Task<()>`
    = note: required because it appears within the type `Option<glommio::Task<()>>`

```

Comment (glommer) [reply src/local_spawn_handle.rs:23]: This seems to refer to Glommio's Task JoinHandle. Because Glommio uses a thread-per-core design, it leverages a lot of optimizations that eliminates atomics in the task's state machine. That means that it can never escape the executor in which it was created. So It indeed cannot be `Send`.

Comment (qiujiangkun) [reply src/local_spawn_handle.rs:23]: So the choices are:
1. Go back to RemoteHandle at cost of performance drop
2. Split `JoinHandle` to `JoinHandle(:Send)` and `LocalJoinHandle(:!Send)` at the cost of a beaking change.

**Comment by qiujiangkun** [comment]
What do you think of the second solution?
```rust
trait LocalSpawnHandle {
fn local_spawn_handle_obj() ->  LocalJoinHandle;
}
trait SpawnHandle {
fn spawn_handle_obj() ->  JoinHandle;
}
struct LocalJoinHandle: !Send {
...
}
impl TryInto<JoinHandle> for LocalJoinHandle {
...
}
struct JoinHandle: Send {
...
}
```

File: src/local_spawn_handle.rs (hunk 6 above)
Comment (najamelan) [reply src/local_spawn_handle.rs:23]: I feel the little bit of perf we would get out of this isn't worth breaking up the API. Also, even when using a local executor, it can be handy and desirable to await the outcome on another thread when `Out` is `Send`. If we make `LocalSpawnHandle` return a `!Send` `JoinHandle`, we not only affect `glommio`, but also other executors which don't have this problem.

So I would say let's do `RemoteHandle` for this. The overhead is just a matter of a oneshot channel.

**Comment by najamelan** [comment]
@glommer thanks for chipping in. I wanted to ask a clarification on how glommio works. What happens if a user instantiates `LocalExecutor` twice in the same thread?

File: src/local_spawn_handle.rs (hunk 6 above)
Comment (qiujiangkun) [reply src/local_spawn_handle.rs:23]: Should all task bahave like "if we drop it without `detach`ing, it will be cancelled"? I think RemoteHandle cannot cancel tasks at least.

Comment (najamelan) [reply src/local_spawn_handle.rs:23]: Yes, all have consistent behavior. That's what most of the code in joinhandle.rs does as tokio and async-std reverse this. `RemoteHandle` already does this out of the box. But all we need is to use `JoinHandle::RemoteHandle`. All the code is already there for detaching and dropping.

File: tests/glommio_ct.rs [hunk 7]
```diff
@@ -1,8 +1,8 @@
 #![ cfg( feature = "glommio" ) ]
 
-// ✔ pass a     ThreadPool  to a function that takes exec: `impl SpawnHandle`
-// ✔ pass a Arc<ThreadPool> to a function that takes exec: `impl SpawnHandle`
-// ✔ pass a    &ThreadPool  to a function that takes exec: `&dyn SpawnHandle`
+// ✔ pass a     GlommioCt  to a function that takes exec: `impl SpawnHandle`
```
Comment (najamelan) [inline tests/glommio_ct.rs:3]: Could you take the file test/tokio_ct.rs instead of the one from `ThreadPool` as a basis? This one doesn't test `LocalSpawn`  and  `LocalSpawnHandle`, as a threadpool does not implement these traits, but for glommio we want them tested.

**Review by najamelan — CHANGES_REQUESTED** [review]
Ok, I think it looks good except for the integration test.

File: tests/glommio_ct.rs (hunk 7 above)
Comment (qiujiangkun) [reply tests/glommio_ct.rs:11]: Ok, now I see what happens if a user instantiates LocalExecutor twice in the same thread. If the memory limit in /etc/security/limits.conf is not high enough, it will report OutOfMemory error
https://github.com/DataDog/glommio/issues/1

**Review by najamelan — APPROVED** [review]
Awesome, it looks great. I'll take it from here. I will go through the code, add a few commits for formatting and then I will merge.

**Comment by glommer** [comment]
Since the executor itself is blocking (equivalent of tokio's block_on), I
don't see how it is possible to create two executors in the same thread.

**Comment by najamelan** [comment]
ok, that makes sense. So in order to spawn futures, one has to start from a block_on function and within that function, I suppose I'll see what happens if I try to make another executor.

**Comment by najamelan** [comment]
@glommer So it turns out it's possible to create an executor and call `block_on` from within a future running inside `block_on` of another `LocalExecutor`. Any idea what happens under the hood? Does it create 2 different executors and it just runs the inner one first, or does it change the configuration of the outer one?

**Comment by najamelan** [comment]
@qiujiangkun is "allow edits by maintainers" checked on the PR? I can't seem to push to HappyCoding/async_executors:glommio branch.

**Comment by qiujiangkun** [comment]
I remembered I checked the box. But now the repo is under an org account, which makes `allow edits by maintainers` not working. As a workaround, you can open a PR to `HappyCoding/async_executors:glommio` and I'll accept it

**Comment by najamelan** [comment]
Ok, I'll just merge it and commit on dev.

**Comment by najamelan** [comment]
Do you have any idea how we increase rlimit_memlock on github actions?

**Comment by najamelan** [comment]
> My main use of it is not `io_uring`, but its thread model. As far as I know, it is the only mature runtime that supports `pin_to_cpu`. It beat tokio by 30% in one of my benchmarks.

Just realized. You should be able to pin the thread rather than the executor ([core_affinity](https://docs.rs/core_affinity/0.5.10/core_affinity/)). This should allow you to pin any single threaded executor.

**Comment by qiujiangkun** [comment]
> > My main use of it is not `io_uring`, but its thread model. As far as I know, it is the only mature runtime that supports `pin_to_cpu`. It beat tokio by 30% in one of my benchmarks.
> 
> Just realized. You should be able to pin the thread rather than the executor ([core_affinity](https://docs.rs/core_affinity/0.5.10/core_affinity/)). This should allow you to pin any single threaded executor.

I know of it, but the problem is that most runtimes don't give you the change to. Take tokio-tp for example, where to you set core_affinity for each thead in the thread_pool?

**Comment by najamelan** [comment]
> Take tokio-tp for example, where to you set core_affinity for each thead in the thread_pool?

Tokio builder let's you initialize the worker threads with [`on_thread_start`](https://docs.rs/tokio/1.6.1/tokio/runtime/struct.Builder.html#method.on_thread_start). From async_executors you can access this builder from [`TokioTpBuilder::tokio_builder`](https://docs.rs/async_executors/0.4.1/async_executors/struct.TokioTpBuilder.html#method.tokio_builder) to configure it.

**Comment by glommer** [comment]
> @glommer So it turns out it's possible to create an executor and call `block_on` from within a future running inside `block_on` of another `LocalExecutor`. Any idea what happens under the hood? Does it create 2 different executors and it just runs the inner one first, or does it change the configuration of the outer one?
> 
> > Since the executor itself is blocking (equivalent of tokio's block_on), I don't see how it is possible to create two executors in the same thread.
> 
> In comparison, tokio `block_on` panics when called from async context.

I think we should panic as well. This is unworkable because our executors rely heavily on TLS variables.
I will add a test and a fix for that.

Thanks for bringing this to my attention!

**Comment by najamelan** [comment]
@glommer. If you are already using TLS and thus a global executor, maybe it's worth providing a way for the user to obtain an `Option<LocalExecutor>`. It would be used here in any case, because then we can make the wrapper ZST. The builder could return `Result<LocalExecutor, LocalExecutor>. If none exist, you get a new one with the configuration you set on the builder, but if one already exists, you get the old one as an error, which means your new configs are ignored.

There is other patterns for this like `get_or_init` API's. A crate like [once_cell](https://crates.io/crates/once_cell) makes this quite convenient. Can't remember if they require `Send + Sync` though.
